src/auth/stscredentials.ts

```diff
@@ -0,0 +1,179 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {GaxiosOptions, GaxiosResponse} from 'gaxios';
+import * as querystring from 'querystring';
+
+import {DefaultTransporter} from '../transporters';
+import {Headers} from './oauth2client';
+import {
+  ClientAuthentication,
+  OAuthClientAuthHandler,
+  OAuthErrorResponse,
+  getErrorFromOAuthErrorResponse,
+} from './oauth2common';
+
+/**
+ * Defines the interface needed to initialize an StsCredentials instance.
+ * The interface does not directly map to the spec and instead is converted
+ * to be compliant with the JavaScript style guide. This is because this is
+ * instantiated internally.
+ * StsCredentials implement the OAuth 2.0 token exchange based on
+ * https://tools.ietf.org/html/rfc8693.
+ * Request options are defined in
+ * https://tools.ietf.org/html/rfc8693#section-2.1
+ */
+export interface StsCredentialsOptions {
+  grantType: string;
+  resource?: string;
+  audience?: string;
+  scope?: string[];
+  requestedTokenType?: string;
+  subjectToken: string;
+  subjectTokenType: string;
+  actingParty?: {
+    actorToken: string;
+    actorTokenType: string;
+  };
+}
+
+/**
+ * Defines the standard request options as defined by the OAuth token
+ * exchange spec: https://tools.ietf.org/html/rfc8693#section-2.1
+ */
+interface StsRequestOptions {
+  grant_type: string;
+  resource?: string;
+  audience?: string;
+  scope?: string;
+  requested_token_type?: string;
+  subject_token: string;
+  subject_token_type: string;
+  actor_token?: string;
+  actor_token_type?: string;
+  client_id?: string;
+  client_secret?: string;
+  [key: string]: string | undefined;
```

Comment:
That is less than ideal :/  It effectively means any string key is value here with a string value.  Is this really an open property bag?

Reply:
Ok. Currently there is only one non-standard field here. I manually added it.